File: src/main/java/app/askresume/api/access/controller/AccessController.java
package app.askresume.api.access.controller;

import app.askresume.api.access.dto.request.LoginRequest;
import app.askresume.api.access.dto.response.AccessTokenResponse;
import app.askresume.api.access.dto.response.LoginResponse;
import app.askresume.api.access.facade.AccessFacade;
import app.askresume.api.access.validator.OauthValidator;
import app.askresume.domain.member.constant.MemberType;
import app.askresume.global.model.ApiResult;
import app.askresume.global.resolver.token.AuthorizationToken;
import app.askresume.global.resolver.token.TokenDto;
import app.askresume.api.access.dto.request.OauthLoginRequest;
import app.askresume.api.access.dto.request.SignUpRequest;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@Tag(name = "authentication", description = "로그인/로그아웃/토큰재발급 API")
@RestController
@RequiredArgsConstructor
@RequestMapping("/api")
public class AccessController {

    private final OauthValidator oauthValidator;

    private final AccessFacade accessFacade;

    @Tag(name = "authentication")
    @Operation(summary = "일반 로그인 API", description = "일반 로그인 API")
    @PostMapping("/login")
    public ResponseEntity<ApiResult<LoginResponse>> login(@RequestBody LoginRequest loginRequest) {


        return ResponseEntity.ok(new ApiResult<>(accessFacade.login(loginRequest)));
    }

    @Tag(name = "authentication")
    @Operation(summary = "일반 회원가입 API", description = "일반 회원가입 API")
    @PostMapping("/sign-up")
    public ResponseEntity<Void> register(@Validated @RequestBody SignUpRequest signUpRequest) {
        accessFacade.register(signUpRequest);
        return ResponseEntity.status(HttpStatus.CREATED).build();
    }


    @Tag(name = "authentication")
    @Operation(summary = "소셜 로그인 API", description = "소셜 로그인 API")
    @PostMapping("/oauth/login")
    public ResponseEntity<ApiResult<LoginResponse>> oauthLogin(@RequestBody OauthLoginRequest oauthLoginRequestDto,
                                                               @AuthorizationToken TokenDto token) {

        oauthValidator.validateMemberType(oauthLoginRequestDto.memberType());
        MemberType memberType = MemberType.from(oauthLoginRequestDto.memberType());

        return ResponseEntity.ok(new ApiResult<>(accessFacade.oauthLogin(token.getToken(), memberType)));
    }

    @Tag(name = "authentication")
    @Operation(summary = "로그아웃 API", description = "로그아웃시 refresh token 만료 처리")
    @PostMapping("/logout")
    public ResponseEntity<Void> logout(@AuthorizationToken TokenDto token) {
        accessFacade.logout(token.getToken());

        return ResponseEntity.noContent().build();
    }

    @Tag(name = "authentication")
    @Operation(summary = "Access Token 재발급 API", description = "Access Token 재발급 API")
    @PostMapping("/reissued/access-token")
    public ResponseEntity<ApiResult<AccessTokenResponse>> createAccessToken(@AuthorizationToken TokenDto token) {
        return ResponseEntity.ok(new ApiResult<>(accessFacade.createAccessTokenByRefreshToken(token.getToken())));
    }
}

File: src/main/kotlin/app/askresume/api/access/controller/AccessController.kt
package app.askresume.api.access.controller

import app.askresume.api.access.usecase.AccessUseCase
import app.askresume.global.cookie.CookieOption
import app.askresume.global.cookie.CookieProvider
import app.askresume.global.jwt.constant.JwtTokenType
import app.askresume.global.resolver.token.AccessToken
import app.askresume.global.resolver.token.RefreshToken
import app.askresume.global.resolver.token.TokenDto
import app.askresume.oauth.OAuthProperties
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.http.HttpHeaders
import org.springframework.http.ResponseCookie
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.time.Duration
import javax.servlet.http.HttpServletRequest

@Tag(name = "authentication", description = "로그인/로그아웃/토큰재발급 API")
@RestController
@RequestMapping("/api")
class AccessController(
    private val accessUseCase: AccessUseCase,
    private val cookieProvider: CookieProvider,
    private val oAuthProperties: OAuthProperties,
) {

    @Tag(name = "authentication")
    @Operation(summary = "로그아웃 API", description = "로그아웃시 refresh token 만료 처리")
    @PostMapping("/logout")
    fun logout(@AccessToken token: TokenDto, request: HttpServletRequest): ResponseEntity<Void> {
        accessUseCase.logout(token.token)

        val expiredCookies = getAccessAndRefreshTokenCookies(oAuthProperties.domain, Duration.ZERO)

        val headers = HttpHeaders()
        expiredCookies.forEach { headers.add(HttpHeaders.SET_COOKIE, it.toString()) }

        return ResponseEntity.noContent().headers(headers).build()
    }

    private fun getAccessAndRefreshTokenCookies(domain: String, maxAge: Duration): Array<ResponseCookie> {
        val expiredAccessTokenCookie = cookieProvider.createCookie(
            CookieOption(name = JwtTokenType.ACCESS.cookieName, domain = domain, maxAge = maxAge))

        val expiredRefreshTokenCookie = cookieProvider.createCookie(
            CookieOption(name = JwtTokenType.REFRESH.cookieName, domain = domain, maxAge = maxAge))

        return arrayOf(expiredAccessTokenCookie, expiredRefreshTokenCookie)

    }

    @Tag(name = "authentication")
    @Operation(summary = "Access Token 재발급 API", description = "쿠키에 저장된 Refresh 토큰을 읽어와 만료된 Access 토큰을 재발급해줍니다.")
    @GetMapping("/refresh")
    fun createAccessToken(@RefreshToken token: TokenDto): ResponseEntity<Void> {
        val accessTokenDto = accessUseCase.createAccessTokenByRefreshToken(token.token)

        val headers = HttpHeaders()
        headers.add(HttpHeaders.SET_COOKIE, cookieProvider.createTokenCookie(accessTokenDto, oAuthProperties.domain).toString())

        return ResponseEntity
            .ok()
            .headers(headers)
            .build()
    }

}

